src/grpc/infer_handler.cc

```diff
@@ -916,8 +916,29 @@ ModelInferHandler::Execute(InferHandler::State* state)
   if (err == nullptr) {
     TRITONSERVER_InferenceTrace* triton_trace = nullptr;
 #ifdef TRITON_ENABLE_TRACING
-    state->trace_ =
-        std::move(trace_manager_->SampleTrace(request.model_name()));
+    TraceStartOptions start_options;
+    InferenceTraceMode mode = TRACE_MODE_TRITON;
+    trace_manager_->GetTraceMode(request.model_name(), &mode);
+    if (mode == TRACE_MODE_OPENTELEMETRY) {
+#ifndef _WIN32
+      const GrpcServerCarrier carrier(state->context_->ctx_.get());
+      auto prop = otel_cntxt::propagation::GlobalTextMapPropagator::
+          GetGlobalPropagator();
+      auto ctxt = opentelemetry::context::Context();
+      start_options.propagated_context = prop->Extract(carrier, ctxt);
+      otel_trace_api::SpanContext span_context =
+          otel_trace_api::GetSpan(start_options.propagated_context)
+              ->GetContext();
+      if (span_context.IsValid()) {
+        start_options.force_sample = true;
+      }
+#else
+      LOG_ERROR << "Unsupported trace mode: "
+                << TraceManager::InferenceTraceModeString(mode);
+#endif
+    }
```

Comment:
(can be follow-up) Could it be moved to a helper?
```cpp
auto start_options = GetTraceStartOptions(state, request);
state->trace_ = std::move(
        trace_manager_->SampleTrace(request.model_name(), start_options));
```

Reply:
Yes, it most likely should